ruff_lsp/server.py

```diff
@@ -679,99 +968,166 @@ async def resolve_code_action(params: CodeAction) -> CodeAction:
         f"{CodeActionKind.SourceOrganizeImports.value}.ruff",
     ):
         # Generate the "Ruff: Organize Imports" edit
-        results = await _fix_document_impl(document, only="I001")
-        params.edit = _create_workspace_edits(document, results)
+        params.edit = await _fix_document_impl(document, only="I001")
 
     elif settings["fixAll"] and params.kind in (
         CodeActionKind.SourceFixAll,
         f"{CodeActionKind.SourceFixAll.value}.ruff",
     ):
         # Generate the "Ruff: Fix All" edit.
-        results = await _fix_document_impl(document)
-        params.edit = _create_workspace_edits(document, results)
+        params.edit = await _fix_document_impl(document)
 
     return params
 
 
 @LSP_SERVER.command("ruff.applyAutofix")
-async def apply_autofix(arguments: tuple[TextDocument]):
+async def apply_autofix(ls: server.LanguageServer, arguments: tuple[TextDocument]):
     uri = arguments[0]["uri"]
-    text_document = LSP_SERVER.workspace.get_text_document(uri)
-    results = await _fix_document_impl(text_document)
-    LSP_SERVER.apply_edit(
-        _create_workspace_edits(text_document, results),
-        "Ruff: Fix all auto-fixable problems",
-    )
+    document = Document.from_uri(uri)
+    workspace_edit = await _fix_document_impl(document)
+    if workspace_edit is None:
+        return
+    ls.apply_edit(workspace_edit, "Ruff: Fix all auto-fixable problems")
 
 
 @LSP_SERVER.command("ruff.applyOrganizeImports")
-async def apply_organize_imports(arguments: tuple[TextDocument]):
+async def apply_organize_imports(
+    ls: server.LanguageServer, arguments: tuple[TextDocument]
+):
     uri = arguments[0]["uri"]
-    text_document = LSP_SERVER.workspace.get_text_document(uri)
-    results = await _fix_document_impl(text_document, only="I001")
-    LSP_SERVER.apply_edit(
-        _create_workspace_edits(text_document, results),
-        "Ruff: Format imports",
-    )
+    document = Document.from_uri(uri)
+    workspace_edit = await _fix_document_impl(document, only="I001")
+    if workspace_edit is None:
+        return
+    ls.apply_edit(workspace_edit, "Ruff: Format imports")
 
 
 @LSP_SERVER.command("ruff.applyFormat")
 async def apply_format(arguments: tuple[TextDocument]):
     uri = arguments[0]["uri"]
-    text_document = LSP_SERVER.workspace.get_text_document(uri)
-    results = await _format_document_impl(text_document)
-    LSP_SERVER.apply_edit(
-        _create_workspace_edits(text_document, results),
-        "Ruff: Format document",
-    )
+    document = Document.from_uri(uri)
+    results = await _run_format_on_document(document)
+    workspace_edit = _result_to_workspace_edit(document, results)
+    if workspace_edit is None:
+        return
+    LSP_SERVER.apply_edit(workspace_edit, "Ruff: Format document")
```

Comment:
Can you walk me through what happens if we receive a notebook cell here? How does that interact with `_run_format_on_document`?

Reply:
The `applyFormat` command works at the Notebook level. 

For a Notebook the `arguments` is going to contain a single entry which will be a `TextDocument` for the _current_ cell (active cell or the cell containing the cursor). We'll construct a `Document` which represents either a Python file or a Notebook.

After the formatting is done, the `_result_to_workspace_edit` will generate a single `TextDocumentEdit` for the Python file or a list of `TextDocumentEdit` for each cell in  a Notebook. These are then a part of `WorkspaceEdit` which are then applied.